Pubsub

slandau3 opened the pull request:
Hello everyone. Apologies if there are mistakes in this PR. First time contributing to a project.

This is a pull request for issue #114, the pubsub api.

I have implemented python functions for all pubsub methods: ls, pub, sub, and peers. I have also included a few test cases (that all pass) for the aforementioned methods. Note that the test case for "peers" does not actually test anything for I can't predict what peers the function will return.

Another note for the sub api. I didn't want to introduce another dependency for a webhook (?) so I simply returned a stream for that function. In the documentation for the pubsub_sub function I made it clear if a user wants to obtain published content they simply need to iterate over the generator (of the stream) that's returned. 

ex:
```python

sub = pubsub_sub('testing')

# publish something to the testing topic
pubsub_pub('testing', 'message')

# will block until stream is closed so this should be run on a separate thread
for word in sub:
    print(word)   # will print anything published to the 'testing' topic. In this case word will be data about 'message'

```

Once again this is my first time contributing to an open source project so I apologize for any mistakes. I ran `tox` for both codestyle and test cases.

alexander255 commented on test/functional/tests.py:
```diff
@@ -807,6 +807,68 @@ def test_bitswap_unwant(self):
+        at the same time for it's difficult
+        to determine if one works without checking the
+        other.
+        """
```
This sentence confused me originally when reading it for the first time: At least to me it sounds like a TODO-entry. Is this supposed to be present-tense (ie: “We test both publishing and subscribing at the same time because …”) by any chance?

alexander255 commented on test/functional/tests.py:
```diff
@@ -807,6 +807,68 @@ def test_bitswap_unwant(self):
+        other.
+        """
+        # the topic that will be published/subscribed to
+        topic = 'testing'
```
Please use an unique channel topic like per session (for instance by using uuid like in `"{}.testing".format(uuid.uuid4())`): As far as I know pubsub messages are supposed to be impliciently propagated across subscribing nodes in a future version and with these generic names there'd be a high chance of random interference from other nodes (and also interferance from other people possibly running these same unit tests) which would cause “random errors in tests” – something that everybody hates to debug.

(Also in the other test below.)

alexander255 commented on test/functional/tests.py:
```diff
@@ -807,6 +807,68 @@ def test_bitswap_unwant(self):
+        expected_return = { 'Strings': [topic] }
+
+        # subscribe to the topic testing
+        # in this test we don't care about the return value
```
Maybe add something like “… but we need to keep it around for the lifetime of this function to ensure that we stay subscribed to the topic long enough” to make it more obvious what is happening?

slandau3 replied:
This reminds me, when requests performs a request to the subscription url, I believe it will remain open. This could waste resources if someone is constantly subscribing and discarding channels. I don't think it'll close itself because it's still technically listening to the channel. What do you recommend doing for this? Wrapping it in a class that has a function which will close the subscription?

alexander255 commented on test/functional/tests.py:
```diff
@@ -577,31 +577,31 @@ def test_dir_make_fill_list_delete(self):
+        # res = self.api.block_put(path)
+        # for key in expected_keys:
+            # self.assertTrue(key in res)
+        # self.assertEqual(res['Key'], expected_block_multihash)
```
Why comment out the block test?

slandau3 replied:
I commented that out when I was testing on my machine. For some reason it was taking a really long time and I forgot to uncomment it. My bad.

alexander255 commented on test/functional/tests.py:
```diff
@@ -807,6 +807,68 @@ def test_bitswap_unwant(self):
+    def setUp(self):
+        self.api = ipfsapi.Client()
+
+    def test_pubsub_pub_sub(self):
```
Any ideas why this test fails on the CI system? (Error is at https://travis-ci.org/ipfs/py-ipfs-api/jobs/371828905#L621.)

alexander255 replied:
Actually the issue may simply be that the versions of `go-ipfs` used by the CI system is too old. Try updating the URL in https://github.com/ipfs/py-ipfs-api/blob/master/.travis.yml#L17 to whatever the URL of the latest Linux-x86_64 version is. The issue should likely be fixed by that.

whereswaldon replied:
We've definitely seen instances of that in the past, so I agree that this is likely the issue.

alexander255 commented on test/functional/tests.py:
```diff
@@ -807,6 +807,68 @@ def test_bitswap_unwant(self):
+        Not sure how to test this since it fully depends
+        on who we're connected to. We may not even have
+        any peers
+        """
```
Indeed :slightly_smiling_face: 
But please call the function anyways and check if its result appears to make sense (a list with some values probably), so that we can at least detect if the function is totally broken for some reason.

slandau3 replied:
Added a call to the function as well as a few assertions

alexander255 commented:
> Hello everyone. Apologies if there are mistakes in this PR. First time contributing to a project.
>
> I have implemented python functions for all pubsub methods: ls, pub, sub, and peers. I have also included a few test cases (that all pass) for the aforementioned methods. Note that the test case for "peers" does not actually test anything for I can't predict what peers the function will return.

Thanks for providing these! They're very welcome!

> Another note for the sub api. I didn't want to introduce another dependency for a webhook (?) so I simply returned a stream for that function. In the documentation for the pubsub_sub function I made it clear if a user wants to obtain published content they simply need to iterate over the generator (of the stream) that's returned.

No problem! The only way to do fully concurrent requests is to launch another thread currently. You pretty much already got the best of the underlying `requests` library with your approach – anything more would require that port to `aiohttp` that I had started some time ago, but never completed due to some hard-to-debug issues arising that I couldn't figure out.

> Once again this is my first time contributing to an open source project so I apologize for any mistakes. I ran `tox` for both codestyle and test cases.

Again, thanks for the work you provided already! :+1:

slandau3 commented on test/functional/tests.py:
```diff
@@ -807,6 +807,68 @@ def test_bitswap_unwant(self):
+        # get the 
+        sub_data = next(sub)
+
+        self.assertEqual(sub_data['data'], expected_data)
```
Didn't know that. Will do!

slandau3 commented:
Thanks for taking a look at it :smile: ! I'll make the changes you recommended as soon as I get a chance

whereswaldon commented:
Thanks for all of your work so far! I agree with @alexander255's comments. This is close to merge-worthy though. Well done!

slandau3 commented:
Thank you so much @whereswaldon and @alexander255! It's been my pleasure and I'm really glad you guys like it! I really like the idea of IPFS and I want to be able to assist in it's development and adoption in any way possible.

All the tests I wrote appear to pass :smile: yet several tests I did not touch are failing. Those tests also fail on the master branch itself

whereswaldon commented:
I decided to resolve the travis merge conflict in favor of 0.4.14 because it's the latest stable release. It seems like it might be a bad idea to run our tests against a release candidate version.

slandau3 commented:
Awesome. All tests pass. Is there anything left to do before merge? I believe all the changes @alexander255 requested have been made

alexander255 commented:
Its kind of unfair that we're unloading our technical depth on your PR @slandau3, but at least some issues are now being ironed out thanks to you. Aside from documenting the need to close the returned generator I have no further objects.

slandau3 commented:
That's quite all right @alexander255! I appreciate it because it helps me get a better understanding of the program. Glad it's resulted in other issues being ironed out too.

I just updated the tests to close the "generator" returned. I also updated the documentation for pubsub_sub to make it clear to the reader that the returned subscription is a context manager/can be manually closed with `.close()`

I'm happy to help!

slandau3 commented:
I'm not sure what the problem is with test_block_get or those metric tests. I don't know if anything I did could have affected them... I doubt it because I only made minor changes since the last time it said "all tests pass"

whereswaldon commented:
I restarted the Python 2.7 build and it succeeded. Seems like we have some nondeterminism in the tests, which is alarming.

alexander255 commented:
The (occasional) errors with `test_block_get` are annoying and indeed non-deterministic, but they do not represent any issues that need fixing as such (at least on our end): They're simply artifacts caused by the shitty network environment for P2P applications in general – strict NAT with no possibility of port forwarding – inside the CI system. Of course, go-IPFS should probably somehow handle this situation better, but it's tricky and the options are very limited. I hope this will be improved cause many people are stuck in this kind a network environment, but it's not something we can fix within the scope of this project.

slandau3 commented:
Just finished making the changes that were requested

slandau3 commented:
Seems I forgot to add the updates client.py file in the latest commit which is why it's failing some tests (for ls). I'll add it once I get home today

alexander255 commented:
@slandau3: I like the idea of exposing a more problem-specific interface for subscription channels as calling `next(sub)` to get a new message feels kinnda weird. :+1:
As you probably noticed,  it's still missing the `.close` method however. You can also drop the `.__next__` method as well and make `.__iter__` just `return self.__sub` – it's shorter and this way `for msg in sub: …` will work again (the current implementation does not return an iterator).

slandau3 commented:
@alexander255 Sorry for the wait. I made the changes you requested and I'm not sure what I was thinking with the original `__iter__`. Not sure why I used `return` instead of `yield` :thinking:

slandau3 commented:
Thanks @alexander255! Glad I could help. I'll keep a lookout for other features I can work on in the future.
